###ENPM 808W Homework 2

Sravya Lenka
ENPM808W DE01

###Import Statements

In [59]:
#import statements
import pandas as pd
from sklearn.linear_model import LinearRegression, Ridge, ElasticNet, RidgeCV, ElasticNetCV
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import GridSearchCV

import numpy as np


In [60]:
from google.colab import drive
drive.mount('/content/drive',force_remount=True)

Mounted at /content/drive


In [61]:
PATH = "/content/drive/MyDrive/ENPM808W_Class/HW2/enpm808w-fall-2022-hw2/"

train_data = pd.read_csv(PATH+"house_train.csv")
test_data = pd.read_csv(PATH+"house_test.csv")

#Understanding Data

In [62]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8973 entries, 0 to 8972
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         8973 non-null   int64  
 1   zip        8973 non-null   int64  
 2   state      8973 non-null   object 
 3   county     8973 non-null   object 
 4   poverty    8973 non-null   float64
 5   price2007  8973 non-null   int64  
 6   price2013  8973 non-null   int64  
dtypes: float64(1), int64(4), object(2)
memory usage: 490.8+ KB


In [63]:
train_data.describe()

id           zip      poverty     price2007     price2013
count   8973.000000   8973.000000  8973.000000  8.973000e+03  8.973000e+03
mean    5016.424830  45752.551098    11.438504  2.880939e+05  2.424867e+05
std     2900.023449  31306.650907     4.520784  2.276980e+05  2.176485e+05
min        0.000000   1001.000000    -1.000000  3.400000e+04  3.520000e+04
25%     2505.000000  19026.000000     8.400000  1.397000e+05  1.223000e+05
50%     5023.000000  37324.000000    11.100000  2.230000e+05  1.788000e+05
75%     7528.000000  76085.000000    14.500000  3.627000e+05  2.848000e+05
max    10035.000000  99709.000000    34.000000  3.467800e+06  4.709500e+06

In [64]:
train_data.head()

id    zip state         county  poverty  price2007  price2013
0   0  10467    NY          bronx     27.1     335200     294000
1   1  11226    NY          kings     21.9     471500     471600
2   2  60640    IL           cook     14.6     254600     174200
3   3  94109    CA  san francisco     10.6     707100     822600
4   4  11375    NY         queens     12.2     636400     681500

In [65]:
test_data.head()

id    zip state      county  poverty  price2007
0   6  32162    FL      marion     13.0     265600
1  13  78572    TX     hidalgo     34.0      79900
2  20  11212    NY       kings     21.9     332000
3  30  37042    TN  montgomery     12.7      98700
4  37  85032    AZ    maricopa     12.9     266100

In [66]:
train_data["state"].value_counts()

CA    967
FL    693
NY    680
PA    561
OH    431
NJ    409
MA    395
NC    323
IL    323
TX    290
MD    279
MI    271
WA    263
TN    246
CO    220
GA    215
CT    212
VA    211
AZ    198
WI    192
OR    178
MN    177
MO    137
SC    131
OK    126
AL     91
NH     89
IA     81
UT     74
AR     73
NV     61
KY     54
IN     51
RI     51
HI     44
DE     27
NE     27
NM     24
MT     21
AK     20
DC     18
WV     13
LA     11
KS      7
WY      5
ND      3
Name: state, dtype: int64

#Question 2

Predict 2013 home prices using state information only. Answer these questions using all of the training data available (WARNING: by default, Rattle partitions data into validation and test splits).

## Model


In [67]:
train_data_q1 = train_data
train_data_q1 = pd.concat([train_data_q1, pd.get_dummies(train_data_q1["state"], drop_first=True, prefix="state")], axis=1)

train_data_q1

id    zip state         county  poverty  price2007  price2013  \
0         0  10467    NY          bronx     27.1     335200     294000   
1         1  11226    NY          kings     21.9     471500     471600   
2         2  60640    IL           cook     14.6     254600     174200   
3         3  94109    CA  san francisco     10.6     707100     822600   
4         4  11375    NY         queens     12.2     636400     681500   
...     ...    ...   ...            ...      ...        ...        ...   
8968  10031  29915    SC       beaufort     10.4     749800     487700   
8969  10032  95728    CA         nevada      8.8     561300     408800   
8970  10033  84781    UT     washington      8.9     261700     218600   
8971  10034   1338    MA       franklin     10.5     193700     192100   
8972  10035  21405    MD   anne arundel      5.0     907800     719200   

      state_AL  state_AR  state_AZ  ...  state_RI  state_SC  state_TN  \
0            0         0         0  ...         0         0         0   
1            0         0         0  ...         0         0         0   
2            0         0         0  ...         0         0         0   
3            0         0         0  ...         0         0         0   
4            0         0         0  ...         0         0         0   
...        ...       ...       ...  ...       ...       ...       ...   
8968         0         0         0  ...         0         1         0   
8969         0         0         0  ...         0         0         0   
8970         0         0         0  ...         0         0         0   
8971         0         0         0  ...         0         0         0   
8972         0         0         0  ...         0         0         0   

      state_TX  state_UT  state_VA  state_WA  state_WI  state_WV  state_WY  
0            0         0         0         0         0         0         0  
1            0         0         0         0         0         0         0  
2            0         0         0         0         0         0         0  
3            0         0         0         0         0         0         0  
4            0         0         0         0         0         0         0  
...        ...       ...       ...       ...       ...       ...       ...  
8968         0         0         0         0         0         0         0  
8969         0         0         0         0         0         0         0  
8970         0         1         0         0         0         0         0  
8971         0         0         0         0         0         0         0  
8972         0         0         0         0         0         0         0  

[8973 rows x 52 columns]

In [68]:
train_data_q1.drop(['id','state','zip', 'county','poverty','price2007'], axis = 1, inplace = True)

train_data_q1

price2013  state_AL  state_AR  state_AZ  state_CA  state_CO  state_CT  \
0        294000         0         0         0         0         0         0   
1        471600         0         0         0         0         0         0   
2        174200         0         0         0         0         0         0   
3        822600         0         0         0         1         0         0   
4        681500         0         0         0         0         0         0   
...         ...       ...       ...       ...       ...       ...       ...   
8968     487700         0         0         0         0         0         0   
8969     408800         0         0         0         1         0         0   
8970     218600         0         0         0         0         0         0   
8971     192100         0         0         0         0         0         0   
8972     719200         0         0         0         0         0         0   

      state_DC  state_DE  state_FL  ...  state_RI  state_SC  state_TN  \
0            0         0         0  ...         0         0         0   
1            0         0         0  ...         0         0         0   
2            0         0         0  ...         0         0         0   
3            0         0         0  ...         0         0         0   
4            0         0         0  ...         0         0         0   
...        ...       ...       ...  ...       ...       ...       ...   
8968         0         0         0  ...         0         1         0   
8969         0         0         0  ...         0         0         0   
8970         0         0         0  ...         0         0         0   
8971         0         0         0  ...         0         0         0   
8972         0         0         0  ...         0         0         0   

      state_TX  state_UT  state_VA  state_WA  state_WI  state_WV  state_WY  
0            0         0         0         0         0         0         0  
1            0         0         0         0         0         0         0  
2            0         0         0         0         0         0         0  
3            0         0         0         0         0         0         0  
4            0         0         0         0         0         0         0  
...        ...       ...       ...       ...       ...       ...       ...  
8968         0         0         0         0         0         0         0  
8969         0         0         0         0         0         0         0  
8970         0         1         0         0         0         0         0  
8971         0         0         0         0         0         0         0  
8972         0         0         0         0         0         0         0  

[8973 rows x 46 columns]

In [69]:
X_train = train_data_q1.drop(['price2013'], axis = 1)
y_train = train_data_q1['price2013']


In [70]:
X_train.columns

Index(['state_AL', 'state_AR', 'state_AZ', 'state_CA', 'state_CO', 'state_CT',
       'state_DC', 'state_DE', 'state_FL', 'state_GA', 'state_HI', 'state_IA',
       'state_IL', 'state_IN', 'state_KS', 'state_KY', 'state_LA', 'state_MA',
       'state_MD', 'state_MI', 'state_MN', 'state_MO', 'state_MT', 'state_NC',
       'state_ND', 'state_NE', 'state_NH', 'state_NJ', 'state_NM', 'state_NV',
       'state_NY', 'state_OH', 'state_OK', 'state_OR', 'state_PA', 'state_RI',
       'state_SC', 'state_TN', 'state_TX', 'state_UT', 'state_VA', 'state_WA',
       'state_WI', 'state_WV', 'state_WY'],
      dtype='object')

In [71]:
model = LinearRegression()
model.fit(X_train,y_train)

print(model.coef_)
print(model.intercept_)


[-1.40892637e+05 -1.59320411e+05 -7.47779798e+04  2.30872896e+05
 -3.31445455e+04  6.70160377e+03  2.32558889e+05 -7.44633333e+04
 -1.06742121e+05 -1.39333721e+05  1.98581364e+05 -1.28573210e+05
 -1.06567771e+05 -1.46506471e+05 -1.43630000e+05 -1.35713333e+05
 -1.44257273e+05  5.77500000e+04  1.15519713e+02 -1.38182030e+05
 -7.54763277e+04 -1.27859197e+05 -5.57347619e+04 -1.32108019e+05
 -1.24863333e+05 -1.42711481e+05 -7.92850562e+04  2.47604645e+04
 -9.23758333e+04 -6.66660656e+04  1.70166176e+04 -1.64368283e+05
 -1.71825238e+05 -6.68951685e+04 -1.13104332e+05 -5.86692157e+04
 -1.35617786e+05 -1.50723089e+05 -1.21074828e+05 -4.37300000e+04
  1.83131280e+04 -1.52661217e+04 -1.18829479e+05 -1.83306923e+05
 -8.72500000e+04]
281729.9999999994


## Questions

a. What is the intercept? What does it correspond to?

b. How do you get this information from your regression?


Intercept is 281729.99 - it represents the price2013 prediction when all other factors are 0. Considering there are dummy categorical variables, and variables like poverty and price 2007 it doesn't imply anything directly.

model.intercept_ gives the value based on the linear regression model.

c. Based on your regression coefficients, what states have the most and least
expensive average homes?

d. How do you get this information from your regression?

In [72]:
print(max(model.coef_))
print(min(model.coef_))

coef_list = model.coef_.tolist()
index_max = coef_list.index(max(model.coef_))
index_min = coef_list.index(min(model.coef_))

print(X_train.columns[index_max])
print(X_train.columns[index_min])


232558.8888888895
-183306.9230769221
state_DC
state_WV


DC has the most expensive average home and West Virginia has the least expensive average homes based on the train data.

model.coef_ gives the value based on the linear regression model.

e. What is the average price of homes in those states?

f. How do you get this information from your regression?

In [73]:
281729.99 -183306.92

98423.06999999998

Intercept is 281729.99, DC coefficient -> 232558.89 and WV coefficient is -183306.92

Average price for DC -> 514288.88

Average price for WV -> 98423.07

#Question 3
Q3
Predict 2013 home prices from state and county information

## Model

In [74]:
train_data_q2 = None
train_data_q2 = train_data

In [75]:
train_data_q2 = pd.concat([train_data_q2, pd.get_dummies(train_data_q2["state"], drop_first=True, prefix="state")], axis=1)
train_data_q2 = pd.concat([train_data_q2, pd.get_dummies(train_data_q2["county"], drop_first=True, prefix="county")], axis=1)

# train_data_q2['state_county'] = train_data_q2['state']+" " +train_data_q2['county']
# train_data_q2 = pd.concat([train_data_q2, pd.get_dummies(train_data_q2["state_county"], drop_first=True, prefix="state_county")], axis=1)
train_data_q2

id    zip state         county  poverty  price2007  price2013  \
0         0  10467    NY          bronx     27.1     335200     294000   
1         1  11226    NY          kings     21.9     471500     471600   
2         2  60640    IL           cook     14.6     254600     174200   
3         3  94109    CA  san francisco     10.6     707100     822600   
4         4  11375    NY         queens     12.2     636400     681500   
...     ...    ...   ...            ...      ...        ...        ...   
8968  10031  29915    SC       beaufort     10.4     749800     487700   
8969  10032  95728    CA         nevada      8.8     561300     408800   
8970  10033  84781    UT     washington      8.9     261700     218600   
8971  10034   1338    MA       franklin     10.5     193700     192100   
8972  10035  21405    MD   anne arundel      5.0     907800     719200   

      state_AL  state_AR  state_AZ  ...  county_yadkin  county_yakima  \
0            0         0         0  ...              0              0   
1            0         0         0  ...              0              0   
2            0         0         0  ...              0              0   
3            0         0         0  ...              0              0   
4            0         0         0  ...              0              0   
...        ...       ...       ...  ...            ...            ...   
8968         0         0         0  ...              0              0   
8969         0         0         0  ...              0              0   
8970         0         0         0  ...              0              0   
8971         0         0         0  ...              0              0   
8972         0         0         0  ...              0              0   

      county_yamhill  county_yates  county_yavapai  county_yellowstone  \
0                  0             0               0                   0   
1                  0             0               0                   0   
2                  0             0               0                   0   
3                  0             0               0                   0   
4                  0             0               0                   0   
...              ...           ...             ...                 ...   
8968               0             0               0                   0   
8969               0             0               0                   0   
8970               0             0               0                   0   
8971               0             0               0                   0   
8972               0             0               0                   0   

      county_yolo  county_york  county_yuba  county_yuma  
0               0            0            0            0  
1               0            0            0            0  
2               0            0            0            0  
3               0            0            0            0  
4               0            0            0            0  
...           ...          ...          ...          ...  
8968            0            0            0            0  
8969            0            0            0            0  
8970            0            0            0            0  
8971            0            0            0            0  
8972            0            0            0            0  

[8973 rows x 679 columns]

In [76]:
X_train = train_data_q2.drop(['id','zip','state','county','poverty','price2007','price2013'], axis = 1)
y_train = train_data_q2['price2013']
X_train.columns

Index(['state_AL', 'state_AR', 'state_AZ', 'state_CA', 'state_CO', 'state_CT',
       'state_DC', 'state_DE', 'state_FL', 'state_GA',
       ...
       'county_yadkin', 'county_yakima', 'county_yamhill', 'county_yates',
       'county_yavapai', 'county_yellowstone', 'county_yolo', 'county_york',
       'county_yuba', 'county_yuma'],
      dtype='object', length=672)

In [77]:
model = LinearRegression()
model.fit(X_train,y_train)

print(model.coef_)
print(model.intercept_)

[ 6.44525684e+16  6.44525684e+16  6.44525684e+16  6.44525684e+16
  6.44525684e+16  6.44525684e+16 -4.33664185e+18  6.44525684e+16
  6.44525684e+16  6.44525684e+16  2.56598997e+17  6.44525684e+16
  6.44525684e+16  6.44525684e+16  6.44525684e+16  6.44525684e+16
  7.76226272e+17  6.44525684e+16  6.44525684e+16  6.44525684e+16
  6.44525684e+16  6.44525684e+16 -5.86477645e+17  6.44525684e+16
 -4.90294955e+18  6.44525684e+16  6.44525684e+16  6.44525684e+16
 -6.44134566e+17  6.44525684e+16  6.44525684e+16  6.44525684e+16
  6.44525684e+16  6.44525684e+16  6.44525684e+16  6.44525684e+16
  6.44525684e+16  6.44525684e+16  6.44525684e+16  6.44525684e+16
  6.44525684e+16  6.44525684e+16  6.44525684e+16  6.44525684e+16
  6.01647180e+17  2.20400000e+04  5.25120000e+04  6.11040000e+04
  1.66073000e+05  4.64600000e+04  1.45874000e+05  3.43362000e+05
  3.12160000e+04 -1.60168000e+05  1.60000000e+04  5.13440000e+04
 -2.28786000e+05  6.44525684e+16  2.41280000e+04  1.11904000e+05
  4.43120000e+04  4.80000

In [78]:
count = 0
scount = 0
county_name_index  = -1
for i,county in enumerate(X_train.columns):
  if(county.startswith("county_")):
    if(county_name_index == -1):
      county_name_index = i
    count+=1
  else:
    scount+=1
print(count, scount)
print(county_name_index)

627 45
45


In [79]:
coef_list = model.coef_.tolist()[county_name_index:]
print(len(coef_list))

print(max(coef_list),min(coef_list))

index_max = coef_list.index(max(coef_list))
index_min = coef_list.index(min(coef_list))

full_coef_list = model.coef_.tolist()
print("coeff is", full_coef_list[45+index_max] )
print("coeff is", full_coef_list[45+index_min] )

print(X_train.columns[45+index_max])
print(X_train.columns[45+index_min])

627
4.967402114523867e+18 -7.117737035922405e+17
coeff is 4.967402114523867e+18
coeff is -7.117737035922405e+17
county_cass
county_east baton rouge


In [80]:
train_data_q2["county"].describe()

count            8973
unique            628
top       los angeles
freq              212
Name: county, dtype: object

In [81]:
county_counts = train_data_q2["county"].value_counts().reset_index(name= "counts")
county_counts.rename(columns={'index': 'county_cc'}, inplace=True)
county_counts


county_cc  counts
0    los angeles     212
1         orange     141
2     montgomery     122
3      middlesex     115
4        suffolk     110
..           ...     ...
623  king george       1
624    edgecombe       1
625      kaufman       1
626     gallatin       1
627         dade       1

[628 rows x 2 columns]

In [82]:
print(county_counts.loc[county_counts['county_cc'] == "cass", 'counts'].values[0])
print(county_counts.loc[county_counts['county_cc'] == "east baton rouge", 'counts'].values[0])


3
8


In [83]:
avg_price_df = train_data_q2.groupby('county')['price2013'].mean().reset_index(name = "mean")
avg_price_df = avg_price_df.sort_values("mean")
avg_price_df

county          mean
180      edgecombe  5.170000e+04
139       crockett  5.660000e+04
245         hardee  5.690000e+04
580          upson  5.950000e+04
90     cattaraugus  6.460000e+04
..             ...           ...
500  san francisco  1.026417e+06
345          marin  1.052213e+06
387      nantucket  1.092500e+06
440         pitkin  1.092700e+06
504      san mateo  1.131755e+06

[628 rows x 2 columns]

In [84]:
merged_df = pd.merge(avg_price_df, county_counts, left_on='county', right_on="county_cc")
merged_df.drop("county_cc", axis = 1, inplace = True)
merged_df = merged_df.sort_values(["counts","mean"], ascending=False)
merged_df

county           mean  counts
617  los angeles  599370.283019     212
593       orange  446802.127660     141
532   montgomery  273166.393443     122
584    middlesex  397404.347826     115
595      suffolk  448259.090909     110
..           ...            ...     ...
13      bradford   77000.000000       1
7     chautauqua   71600.000000       1
5         hendry   64900.000000       1
3          upson   59500.000000       1
0      edgecombe   51700.000000       1

[628 rows x 3 columns]

In [85]:
merged_df.describe()

mean      counts
count  6.280000e+02  628.000000
mean   1.917605e+05   14.288217
std    1.339377e+05   18.806444
min    5.170000e+04    1.000000
25%    1.180075e+05    4.000000
50%    1.528296e+05    8.000000
75%    2.174267e+05   17.250000
max    1.131755e+06  212.000000

In [86]:
counties = ["cass","east baton rouge"]
merged_df[merged_df["county"].isin(counties)]


county           mean  counts
212  east baton rouge  128700.000000       8
325              cass  156866.666667       3

## Questions

*a*. What US counties have the highest and lowest regression coefficients? Why?

County Cass has the highest regression coefficient (4.967402114523867e+18).

County East Baton Rouge has the lowest regression coefficent (-7.117737035922405e+17).

The regression coefficients indicate the price relative to the parameters of the model i.e. prices based on state and county-related parameters, these coefficients signify the change in price associated with a one-unit change in the relevant state and county variables.

Based on my understanding and since the model knows information about only county, state and prices, I believe County Cass has the highest regression coefficient because it has a mean price greater than average of the whole train data even though it has less count(sample size). This suggests that County Cass possesses unique characteristics contributing to higher prices, requiring further investigation into what factors might be at play.

While County East Baton Rouge has only 8 rows of data but the average price falls near the 25th quartile. But due to the small sample size, it might suggest lower average prices, but the limited data points raise questions about the reliability of this estimate.



#Question 4
Challenge: Build a regressor that best predicts average home values in this dataset. Upload your predictions to Kaggle (link to be provided). (IMPORTANT: you must use your UMD e-mail to get access to this competition). If your UMD username is JSMITH, use JSMITH_DID as your username on Kaggle. You are welcome to use any additional data you care to use so long as they are from 2007 or earlier. You must describe any additional data you use in the writeup. All students must make a submission to Kaggle; additional points available to those who do best (WARNING: make sure you submit something before the deadline, as you cannot use late days on the competition).
a. Describe what you did to build the best predictor possible
b. Give your best Kaggle score
c. Give your Kaggle username

In [87]:
train_data_q3 = train_data
train_data_q3['state_county'] = train_data_q3['state']+" " +train_data_q3['county']
train_data_q3 = pd.concat([train_data_q3, pd.get_dummies(train_data_q3["state_county"], drop_first=True, prefix="state_county")], axis=1)
# train_data_q3 = pd.concat([train_data_q3, pd.get_dummies(train_data_q3["zip"], drop_first=True, prefix="zip")], axis=1)

test_data_q3 = test_data
test_data_q3['state_county'] = test_data_q3['state']+" " + test_data_q3['county']
test_data_q3 = pd.concat([test_data_q3, pd.get_dummies(test_data_q3["state_county"], drop_first=True, prefix="state_county")], axis=1)
# test_data_q3 = pd.concat([test_data_q3, pd.get_dummies(test_data_q3["zip"], drop_first=True, prefix="zip")], axis=1)

train_data_q3

# train_data_q3 = pd.concat([train_data_q3, pd.get_dummies(train_data_q3["state"], drop_first=True, prefix="state")], axis=1)
# train_data_q3 = pd.concat([train_data_q3, pd.get_dummies(train_data_q3["county"], drop_first=True, prefix="county")], axis=1)
# test_data_q3 = pd.concat([test_data_q3, pd.get_dummies(test_data_q3["state"], drop_first=True, prefix="state")], axis=1)
# test_data_q3 = pd.concat([test_data_q3, pd.get_dummies(test_data_q3["county"], drop_first=True, prefix="county")], axis=1)


#adding average price to county

# avg_price_df = train_data_q3.groupby(['state'])['price2007'].mean().reset_index(name = "price_mean_2007")


# train_data_q3 = pd.merge(train_data_q3, avg_price_df, on=['state'], how='left')

# train_data_q3

# avg_price_df_test = test_data_q3.groupby(['state'])['price2007'].mean().reset_index(name = "price_mean_2007")


# test_data_q3 = pd.merge(test_data_q3, avg_price_df_test, on=['state'], how='left')

# avg_price_df


id    zip state         county  poverty  price2007  price2013  \
0         0  10467    NY          bronx     27.1     335200     294000   
1         1  11226    NY          kings     21.9     471500     471600   
2         2  60640    IL           cook     14.6     254600     174200   
3         3  94109    CA  san francisco     10.6     707100     822600   
4         4  11375    NY         queens     12.2     636400     681500   
...     ...    ...   ...            ...      ...        ...        ...   
8968  10031  29915    SC       beaufort     10.4     749800     487700   
8969  10032  95728    CA         nevada      8.8     561300     408800   
8970  10033  84781    UT     washington      8.9     261700     218600   
8971  10034   1338    MA       franklin     10.5     193700     192100   
8972  10035  21405    MD   anne arundel      5.0     907800     719200   

          state_county  state_county_AK fairbanks north star  \
0             NY bronx                                     0   
1             NY kings                                     0   
2              IL cook                                     0   
3     CA san francisco                                     0   
4            NY queens                                     0   
...                ...                                   ...   
8968       SC beaufort                                     0   
8969         CA nevada                                     0   
8970     UT washington                                     0   
8971       MA franklin                                     0   
8972   MD anne arundel                                     0   

      state_county_AK kenai peninsula  ...  state_county_WI saint croix  \
0                                   0  ...                            0   
1                                   0  ...                            0   
2                                   0  ...                            0   
3                                   0  ...                            0   
4                                   0  ...                            0   
...                               ...  ...                          ...   
8968                                0  ...                            0   
8969                                0  ...                            0   
8970                                0  ...                            0   
8971                                0  ...                            0   
8972                                0  ...                            0   

      state_county_WI sheboygan  state_county_WI walworth  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   
...                         ...                       ...   
8968                          0                         0   
8969                          0                         0   
8970                          0                         0   
8971                          0                         0   
8972                          0                         0   

      state_county_WI washington  state_county_WI waukesha  \
0                              0                         0   
1                              0                         0   
2                              0                         0   
3                              0                         0   
4                              0                         0   
...                          ...                       ...   
8968                           0                         0   
8969                           0                         0   
8970                           0                         0   
8971                           0                         0   
8972                           0   

In [88]:
#ValueError: The feature names should match those that were passed during fit.

train_columns = train_data_q3.columns
test_columns = test_data_q3.columns


for column in train_columns:
    if column not in test_data_q3.columns:
        test_data_q3[column] = 0
for column in test_columns:
    if column not in train_data_q3.columns:
        train_data_q3[column] = 0

desired_order = train_data_q3.columns

# Sort the columns of df1 based on the order in df2
test_data_q3 = test_data_q3[desired_order]
test_data_q3

<ipython-input-88-eb736cb0ca2f>:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  test_data_q3[column] = 0
<ipython-input-88-eb736cb0ca2f>:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  test_data_q3[column] = 0
<ipython-input-88-eb736cb0ca2f>:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  test_dat

id    zip state      county  poverty  price2007  price2013  \
0         6  32162    FL      marion     13.0     265600          0   
1        13  78572    TX     hidalgo     34.0      79900          0   
2        20  11212    NY       kings     21.9     332000          0   
3        30  37042    TN  montgomery     12.7      98700          0   
4        37  85032    AZ    maricopa     12.9     266100          0   
...     ...    ...   ...         ...      ...        ...        ...   
1058   9972   6282    CT     windham      8.5     242000          0   
1059   9991  60512    IL     kendall      3.7     233400          0   
1060   9993  21912    MD       cecil      9.3     336600          0   
1061   9996  21662    MD      talbot      8.3     772300          0   
1062  10029  80532    CO     larimer     12.2     234000          0   

       state_county  state_county_AK fairbanks north star  \
0         FL marion                                     0   
1        TX hidalgo                                     0   
2          NY kings                                     0   
3     TN montgomery                                     0   
4       AZ maricopa                                     0   
...             ...                                   ...   
1058     CT windham                                     0   
1059     IL kendall                                     0   
1060       MD cecil                                     0   
1061      MD talbot                                     0   
1062     CO larimer                                     0   

      state_county_AK kenai peninsula  ...  state_county_FL washington  \
0                                   0  ...                           0   
1                                   0  ...                           0   
2                                   0  ...                           0   
3                                   0  ...                           0   
4                                   0  ...                           0   
...                               ...  ...                         ...   
1058                                0  ...                           0   
1059                                0  ...                           0   
1060                                0  ...                           0   
1061                                0  ...                           0   
1062                                0  ...                           0   

      state_county_GA union  state_county_LA bossier  state_county_NH grafton  \
0                         0                        0                        0   
1                         0                        0                        0   
2                         0                        0                        0   
3                         0                        0                        0   
4                         0                        0                        0   
...                     ...                      ...                      ...   
1058                      0                        0                        0   
1059                      0                        0                        0   
1060                      0                        0                        0   
1061                      0                        0                        0   
1062                      0                        0                        0   

      state_county_TX hill  state_county_TX hunt  state_county_TX orange  \
0                        0                     0                       0   
1                        0                     0                       0   
2                        0                     0                       0   
3                        0                     0                       0   
4                        0                     0                       0   
...                    ...                   ...                     ...   
1058                     0   

In [89]:
X_train = train_data_q3.drop(['state','county','state_county','price2013'],axis = 1)
y_train = train_data_q3['price2013']

X_test = test_data_q3.drop(['state','state_county','county','price2013'],axis = 1)
# y_test = test_data_q3['price2007']

In [90]:
print("Training Features:", X_train.columns)

# During prediction
print("Prediction Features:", X_test.columns)

Training Features: Index(['id', 'zip', 'poverty', 'price2007',
       'state_county_AK fairbanks north star',
       'state_county_AK kenai peninsula', 'state_county_AK matanuska susitna',
       'state_county_AL baldwin', 'state_county_AL jefferson',
       'state_county_AL marshall',
       ...
       'state_county_FL washington', 'state_county_GA union',
       'state_county_LA bossier', 'state_county_NH grafton',
       'state_county_TX hill', 'state_county_TX hunt',
       'state_county_TX orange', 'state_county_VA carroll',
       'state_county_VA harrisonburg city', 'state_county_WA douglas'],
      dtype='object', length=804)
Prediction Features: Index(['id', 'zip', 'poverty', 'price2007',
       'state_county_AK fairbanks north star',
       'state_county_AK kenai peninsula', 'state_county_AK matanuska susitna',
       'state_county_AL baldwin', 'state_county_AL jefferson',
       'state_county_AL marshall',
       ...
       'state_county_FL washington', 'state_county_GA unio

Linear Regression

In [91]:

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

Ridge Regression

In [92]:
ridge_reg = Ridge(alpha=0.5)
ridge_reg.fit(X_train, y_train)

y_pred = ridge_reg.predict(X_test)

In [93]:
from sklearn.preprocessing import StandardScaler


scaler_X = StandardScaler()
scaler_y = StandardScaler()


numeric_column_names = ["poverty"]
numeric_data_train = X_train[numeric_column_names]
numeric_scaled = scaler_X.fit_transform(numeric_data_train)

numeric_scaled_df = pd.DataFrame(numeric_scaled, columns=numeric_column_names)

X_scaled = pd.concat([numeric_scaled_df, X_train.drop(columns=numeric_column_names)], axis=1)


y_train_array = np.array(y_train)
y_train_2d = y_train_array.reshape(-1, 1)

y = scaler_y.fit_transform(y_train_2d)

regressor = Ridge(alpha=0.5)
regressor.fit(X_scaled,y)


numeric_data_test = X_test[numeric_column_names]
numeric_scaled = scaler_X.fit_transform(numeric_data_test)

numeric_scaled_df = pd.DataFrame(numeric_scaled, columns=numeric_column_names)

X_test_scaled = pd.concat([numeric_scaled_df, X_test.drop(columns=numeric_column_names)], axis=1)
y_pred = regressor.predict(X_test_scaled)
y_pred_original_scale = scaler_y.inverse_transform(y_pred.reshape(-1, 1))
y_pred = y_pred_original_scale.flatten()


Elastic Net regression

In [94]:
model = ElasticNet(alpha=0.5, l1_ratio=0.5)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [95]:
ratios = np.arange(0, 1, 0.01)
alphas = [0.5, 1.0, 10.0]
model = ElasticNetCV(l1_ratio=ratios, alphas=alphas, n_jobs=-1)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:617: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:617: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:617: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 14734149957988.586, tolerance: 34137556279.120075
  model = cd_fast.enet_coordinate_descent_gram(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:617: UserWarning: Coordinate descent 

In [96]:
#Submission

kaggle_submission = pd.DataFrame({'id': test_data['id'], 'prediction': y_pred})

# Save the predictions to a CSV file
kaggle_submission.to_csv('house_predictions.csv', index=False)

##Questions

a. Describe what you did to build the best predictor possible

b. Give your best Kaggle score

c. Give your Kaggle username

a. I used one-hot encoding to make the categorical variables - state and county to convert into numeric variables. I used combination of zip, one hot encoded state and county variables, poverty and price2007 to train the model. I initially tried different models(Linear, Ridge, Elastic Net, Decision Tree) but I settled Ridge Model since I am dealing with multitude of features which I assumed would be better for the predictions. After trying different alpha values, the best model was with 0.5, so I used that. Because we are dealing with prices, I also took mean price for state and county and trained the model with that to have better predictions but that didn't help with the predictions by much.

b. 49738.02419

c. sravyal96 (Sravya Lenka)

#Question 5

5. (10 points) Suppose you have 2 bags.
 Bag #1 has 1 black ball and 2 white balls.
 Bag #2 has 1 black ball and 3 white balls.
 Suppose you pick a bag at random, and select a ball from that bag. What is the probability of selecting a white ball?

P(W)=P(W∣B1)⋅P(B1) + P(W∣B2⋅P(B2)

P(W) - Picking a white ball
P(B1) - Probability of choosing bag 1
P(B2) - Probability of choosing bag 2

P(W∣B1) - Probability of choosing white ball given bag 1
P(W∣B2) - Probability of choosing white ball given bag 2

Picking a bag P(B1) = P(B2) -> ½

Picking a white ball from bag 1 P(W∣B1) - > ⅔

Picking a white ball from bag 2 P(W∣B2) -> ¾

Probability of picking a white ball P(W) -> (½)(⅔) + (½)(¾) = 17/24


#Question 6
6. (10 points) A soccer team wins 60% of its games when it scores the first goal, and 10% of its games when the opposing team scores first. If the team scores the first goal about 30% of the time, what fraction of the games does it win?



100 games -> 60 times when it scores the first goal and 10 times when the opposing team scores

Team scores first about 30 times -> that implies the other 70 times opposite team scores first

P(W)=P(W∣F)⋅P(F)+P(W∣F\`)*P(F\`)

P(W): Probability of winning a game.

P(F): Probability of scoring the first goal = 30/100

P(W∣S): Probability of winning given that the team scored the first goal = 60/100

P(F\`): Probability of opposite team scoring the first goal = 70/100

P(W∣F\`):  Probability of winning given that the opposite team scored the first goal = 10/100

So, based on formula ->

P(W) = (30/100 * 60/100) + (70/100 * 10/100) = 25/100 = 1/4
